pkg/codegen/python/gen.go

```diff
@@ -965,34 +982,75 @@ func (mod *modContext) genResource(res *schema.Resource) (string, error) {
 
 	if !res.IsProvider && res.DeprecationMessage != "" && mod.compatibility != kubernetes20 {
 		escaped := strings.ReplaceAll(res.DeprecationMessage, `"`, `\"`)
-		fmt.Fprintf(w, "warnings.warn(\"\"\"%s\"\"\", DeprecationWarning)\n\n", escaped)
+		fmt.Fprintf(w, "warnings.warn(\"\"\"%s\"\"\", DeprecationWarning)\n\n\n", escaped)
 	}
 
 	// Produce a class definition with optional """ comment.
-	fmt.Fprint(w, "\n")
 	fmt.Fprintf(w, "class %s(%s):\n", name, baseType)
 	if res.DeprecationMessage != "" && mod.compatibility != kubernetes20 {
 		escaped := strings.ReplaceAll(res.DeprecationMessage, `"`, `\"`)
 		fmt.Fprintf(w, "    warnings.warn(\"\"\"%s\"\"\", DeprecationWarning)\n\n", escaped)
 	}
-	// Now generate an initializer with arguments for all input properties.
-	fmt.Fprintf(w, "    def __init__(__self__,\n")
-	fmt.Fprintf(w, "                 resource_name: str,\n")
-	fmt.Fprintf(w, "                 opts: Optional[pulumi.ResourceOptions] = None")
 
-	// If there's an argument type, emit it.
+	// Determine if all inputs are optional.
+	allOptionalInputs := true
 	for _, prop := range res.InputProperties {
-		wrapInput := !prop.IsPlain
-		ty := mod.typeString(prop.Type, true, wrapInput, true /*optional*/, true /*acceptMapping*/)
-		fmt.Fprintf(w, ",\n                 %s: %s = None", InitParamName(prop.Name), ty)
+		allOptionalInputs = allOptionalInputs && !prop.IsRequired
 	}
 
-	// Old versions of TFGen emitted parameters named __name__ and __opts__. In order to preserve backwards
-	// compatibility, we still emit them, but we don't emit documentation for them.
-	fmt.Fprintf(w, ",\n                 __props__=None")
-	fmt.Fprintf(w, ",\n                 __name__=None")
-	fmt.Fprintf(w, ",\n                 __opts__=None):\n")
-	mod.genInitDocstring(w, res)
+	// Now generate __init__ overloads along with an implementation...
+
+	// First, generate an __init__ overload that accepts the resource's inputs from the args class.
```

Comment:
Do we have a preference for which overload shows up first in code completion lists? VS Code shows the overloads in the order they're declared (at least in this case).

<img width="629" alt="Screen Shot 2021-03-15 at 12 59 03 PM" src="https://user-images.githubusercontent.com/710598/111213754-57cf8480-858e-11eb-936c-8d4f4c815822.png">

Reply:
I'd propose having the keyword-based style first ahead of the args-based given that's what people are currently using. We could choose to flip this order for 3.0 to encourage the args-based as the 'default'

Reply:
SGTM. I've pushed a commit that does that.

<img width="618" alt="Screen Shot 2021-03-16 at 6 22 18 AM" src="https://user-images.githubusercontent.com/710598/111315809-f821b980-861f-11eb-979c-e94e0548a915.png">

<img width="619" alt="Screen Shot 2021-03-16 at 6 22 41 AM" src="https://user-images.githubusercontent.com/710598/111315880-066fd580-8620-11eb-9801-909cadd0a43b.png">